legion/engine/physics/mesh_splitter_utils/mesh_half_edge.hpp

```diff
@@ -0,0 +1,268 @@
+#pragma once
+#include <core/core.hpp>
+#include <physics/mesh_splitter_utils/splittable_polygon.h>
+#include <physics/physics_statics.hpp>
+
+namespace legion::physics
+{
+    struct MeshHalfEdge : std::enable_shared_from_this<MeshHalfEdge>
+    {
+        math::vec3 position;
+        math::vec2 uv;
+
+        std::shared_ptr<MeshHalfEdge> nextEdge = nullptr;
+        std::shared_ptr<MeshHalfEdge> pairingEdge = nullptr;
+        std::shared_ptr<MeshHalfEdge> shadowEdge = nullptr;
+
+        std::weak_ptr<SplittablePolygon> owner;
+
+        //TODO: these bools should probably be stored as an array in SplittablePolygon
+        bool isVisited = false;
+        bool isBoundary = false;
+
+        //std::string debugID = "";
+
+        MeshHalfEdge(const math::vec3& pPosition, const math::vec2& pUVs, std::weak_ptr<SplittablePolygon> pOwner) : position(pPosition), uv(pUVs),owner(pOwner)
+        {
+
+        }
+
+        MeshHalfEdge(const math::vec3& pPosition,const math::vec2& pUVs) : position(pPosition),uv(pUVs)
+        {
+
+        }
+
+        MeshHalfEdge(const math::vec3& pPosition) : position(pPosition)
+        {
+
+        }
+
+        auto GetTriangle() 
+        {
+            return std::make_tuple(shared_from_this(), nextEdge, nextEdge->nextEdge);
+        }
+
+        auto GetShadowTriangle()
+        {
+            return std::make_tuple(shadowEdge, nextEdge->shadowEdge, nextEdge->nextEdge->shadowEdge);
+        }
```

Comment:
What happens when there is no triangle? since a HalfEdgeFace can have more than 3 sides

Reply:
in this case, all our HalfEdgeFaces are connected into a triangle